sda_download/sda_download.go

```diff
@@ -0,0 +1,257 @@
+	config, err := helpers.GetAuth(*configPath)
+	if err != nil {
+		return err
+	}
+
+	// Check if the token has expired
+	err = helpers.CheckTokenExpiration(config.AccessToken)
+	if err != nil {
+		return err
+	}
+
+	// Loop through the files and download them
+	for _, filePath := range files {
+		fileIDURL, err := getFileIDURL(*url, config.AccessToken, *datasetID, filePath)
+		if err != nil {
+			return err
+		}
+
+		// Check if the file path contains a userID and if it does,
+		// do not keep it in the file path
+		filePathSplit := strings.Split(filePath, "/")
+		if strings.Contains(filePathSplit[0], "_") {
+			userIDSplit := strings.Split(filePathSplit[0], "_")
+			userSHA := userIDSplit[0]
+			userSHARegex := regexp.MustCompile("^[a-f0-9]{40}$")
+			match := userSHARegex.MatchString(userSHA)
+			if match {
+				filePath = strings.Join(filePathSplit[1:], "/")
+			}
+		}
+
+		outFilename := filePath
+		if *outDir != "" {
+			outFilename = *outDir + "/" + filePath
+		}
+
+		err = downloadFile(fileIDURL, config.AccessToken, outFilename)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// downloadFile downloads the file by using the download URL
+func downloadFile(uri, token, filePath string) error {
+	filePath = strings.TrimSuffix(filePath, ".c4gh")
+	// Get the file body
+	body, err := getResponseBody(uri, token)
+	if err != nil {
+		return fmt.Errorf("failed to get file for download, reason: %v", err)
+	}
+
+	// Create the directory if it does not exist
+	fileDir := filepath.Dir(filePath)
+	err = os.MkdirAll(fileDir, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to create directory, reason: %v", err)
+	}
+
+	outfile, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to create file, reason: %v", err)
+	}
+	defer outfile.Close()
+
+	// Create a new progress container
+	p := mpb.New()
+
+	// Create a new progress bar with the length of the body
+	bar := p.AddBar(int64(len(body)),
+		mpb.PrependDecorators(
+			decor.CountersKibiByte("% .2f / % .2f"),
+		),
+	)
+
+	// Create a proxy reader
+	reader := strings.NewReader(string(body))
+	proxyReader := bar.ProxyReader(reader)
+
+	fmt.Printf("Downloading file to %s\n", filePath)
+	// Copy from the proxy reader (which updates the progress bar) to the file
+	_, err = io.Copy(outfile, proxyReader)
+	if err != nil {
+		return fmt.Errorf("failed to write file, reason: %v", err)
+	}
+
+	// Wait for the progress bar to finish
+	p.Wait()
+
+	return nil
+}
+
+// getFileIDURL gets the datset files, parses the JSON response to get the file ID
+// and returns the download URL for the file
+func getFileIDURL(baseURL, token, dataset, filename string) (string, error) {
+	// Sanitize the base_url
+	baseURL = strings.TrimSuffix(baseURL, "/")
+	if !strings.HasPrefix(baseURL, "http") {
+		return "", fmt.Errorf("invalid URL, missing protocol (http/https)")
+	}
+
+	// Make the url for listing files
+	filesURL := baseURL + "/metadata/datasets/" + dataset + "/files"
```

Comment:
```suggestion
	filesURL, _ := url.JoinPath(baseURL + "/metadata/datasets/" + dataset + "/files")
```

The `url.JoinPath` take care of any multiple slashes and it is safe to ignore the error since we do `url.ParceRequestURI` above.

Reply:
That had to change back because with this way the url.JoinPath was giving the wrong filesURL, it was removing one slash from the dataset which was `https://doi.example/ty009.sfrrss/600.45asasga`. As a result the filesURl was `http://localhost:8080/metadata/datasets/https:/doi.example/ty009.sfrrss/600.45asasga/files` and of course that was not correct

Reply:
Did you add a test case with the funky doi URL as dataset name so we keep testing for it?